File: Core/Workspace/CSharp/Adapters/Allors.Workspace.Adapters.Local/Session/OriginState/WorkspaceOriginState.cs
// <copyright file="Object.cs" company="Allors bvba">
// Copyright (c) Allors bvba. All rights reserved.
// Licensed under the LGPL license. See LICENSE file in the project root for full license information.
// </copyright>

namespace Allors.Workspace.Adapters.Local
{
    internal sealed class WorkspaceOriginState : RecordBasedOriginState
    {
        internal WorkspaceOriginState(Strategy strategy, WorkspaceRecord record) : base(strategy)
        {
            this.WorkspaceRecord = record;
            this.PreviousRecord = this.WorkspaceRecord;
        }

        protected override IRecord Record => this.WorkspaceRecord;

        private WorkspaceRecord WorkspaceRecord { get; set; }

        protected override void OnChange() => this.Strategy.Session.OnChange(this);

        private bool HasChanges => this.ChangedRoleByRelationType?.Count > 0;

        internal void Push()
        {
            if (this.HasChanges)
            {
                this.Workspace.Push(this.Id, this.Class, this.Record?.Version ?? 0, this.ChangedRoleByRelationType);
            }

            this.Reset();
        }

        internal void Reset()
        {
            this.WorkspaceRecord = this.Workspace.GetRecord(this.Id);
            this.ChangedRoleByRelationType = null;
        }
    }
}

File: typescript/libs/workspace/adapters/system/src/lib/session/originstate/WorkspaceOriginState.ts
import { IWorkspaceResult } from '@allors/workspace/domain/system';
import { RoleType } from '@allors/workspace/meta/system';
import { IRecord } from '../../IRecord';
import { UnknownVersion } from '../../Version';
import { WorkspaceRecord } from '../../workspace/WorkspaceRecord';
import { WorkspaceResult } from '../../workspace/WorkspaceResult';
import { Strategy } from '../Strategy';
import { RecordBasedOriginState } from './RecordBasedOriginState';

export class WorkspaceOriginState extends RecordBasedOriginState {
  constructor(public strategy: Strategy, private workspaceRecord: WorkspaceRecord) {
    super();
    this.previousRecord = this.workspaceRecord;
  }

  get roleTypes(): Set<RoleType> {
    return this.class.workspaceOriginRoleTypes;
  }

  get record(): IRecord {
    return this.workspaceRecord;
  }

  get version(): number {
    return this.workspaceRecord.version;
  }

  protected onChange() {
    this.strategy.session.changeSetTracker.onWorkspaceChanged(this);
    this.strategy.session.pushToWorkspaceTracker.onChanged(this);
  }

  push() {
    if (this.hasChanges) {
      this.workspace.push(this.id, this.class, this.record?.version ?? UnknownVersion, this.changedRoleByRelationType);
    }

    this.workspaceRecord = this.workspace.getRecord(this.id);
    this.changedRoleByRelationType = null;
  }

  onPulled(result: WorkspaceResult) {
    const newRecord = this.workspace.getRecord(this.id);
    if (!this.canMerge(newRecord)) {
      result.addMergeError(this.strategy.object);
      return;
    }

    this.workspaceRecord = newRecord;
    this.changedRoleByRelationType = null;
  }
}
